src/circuit/circuit_frame.rs

```diff
@@ -4698,51 +4721,64 @@ pub fn to_unsigned_int<F: LurkField, CS: ConstraintSystem<F>>(
+        &u_decomp,
+    )?;
+    let r_bits = match size {
+        UnsignedType::U32 => &field_bits[0..32],
+        UnsignedType::U64 => &field_bits[0..64],
+    };
+    enforce_pack(
+        &mut cs.namespace(|| "enforce unsigned pack"),
+        r_bits,
+        r_num.clone(),
     )?;
-    enforce_true(&mut cs.namespace(|| "enforce decomposition"), &u_decomp)?;
 
-    // enforce remainder range
-    let r_bits = r_num.to_bits_le(&mut cs.namespace(|| "u64 remainder bit decomp"))?;
-    let output = match size {
-        UnsignedType::U32 => {
-            enforce_at_most_n_bits(&mut cs.namespace(|| "remainder u32 range"), g, r_bits, 32)?;
-            let output_alloc = AllocatedNum::alloc(&mut cs.namespace(|| "to_32"), || {
-                Ok(F::from_u32(v.to_u32_unchecked()).unwrap())
-            })?;
-            let u32_is_remainder = alloc_equal(
-                &mut cs.namespace(|| "check u32 value is the remainder"),
-                &r_num,
-                &output_alloc,
-            )?;
-            enforce_true(
-                &mut cs.namespace(|| "enforce u32 decomposition output"),
-                &u32_is_remainder,
-            )?;
-            output_alloc
-        }
-        UnsignedType::U64 => {
-            enforce_at_most_n_bits(&mut cs.namespace(|| "remainder u64 range"), g, r_bits, 64)?;
-            let output_alloc = AllocatedNum::alloc(&mut cs.namespace(|| "to_64"), || {
-                Ok(F::from_u64(v.to_u64_unchecked()).unwrap())
-            })?;
-            let u64_is_remainder = alloc_equal(
-                &mut cs.namespace(|| "check u64 value is the remainder"),
-                &r_num,
-                &output_alloc,
-            )?;
-            enforce_true(
-                &mut cs.namespace(|| "enforce u64 decomposition output"),
-                &u64_is_remainder,
-            )?;
-            output_alloc
-        }
+    Ok(r_num)
+}
+
+// Convert from num to unsigned integers by taking the least significant bits.
+// The output is a pair of allocated numbers, where the first one corresponds to
+// the u32 coercion, while the second corresponds to the u64 coercion.
+pub fn to_unsigned_integers<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    g: &GlobalAllocations<F>,
+    maybe_unsigned: &AllocatedNum<F>,
+) -> Result<(AllocatedNum<F>, AllocatedNum<F>), SynthesisError> {
+    let v = match maybe_unsigned.get_value() {
+        Some(v) => v,
+        None => F::zero(), //dummy
     };
-    Ok(output)
+    let field_bn = BigUint::from_bytes_le(v.to_repr().as_ref());
+    // Since bit decomposition is expensive, we compute it only once here
+    let field_bits = maybe_unsigned.to_bits_le(&mut cs.namespace(|| "field element bit decomp"))?;
+
+    let r32_num = to_unsigned_integer_helper(
+        &mut cs.namespace(|| "enforce u32"),
+        g,
+        maybe_unsigned,
+        field_bn.clone(),
+        field_bits.clone(),
```

Comment:
Can you refactor to allow passing `&field_bits`?

Reply:
done